watertap/unit_models/gac.py

```diff
@@ -760,209 +708,303 @@ def eq_dg(b, t, j):
+
+        @self.Constraint(doc="relating velocities based on bed voidage")
+        def eq_velocity_relation(b):
+            return b.velocity_sup == b.velocity_int * b.bed_voidage
+
+        @self.Constraint(doc="bed length based on velocity and ebct")
+        def eq_bed_length(b):
+            return b.bed_length == b.velocity_sup * b.ebct
+
+        @self.Constraint(doc="bed diameter and area relation")
+        def eq_bed_diameter(b):
+            return Constants.pi * (b.bed_diameter**2) == 4 * b.bed_area
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc="bed area based on velocity and volumetric flow",
+        )
+        def eq_bed_area(b, t):
+            return (
+                b.bed_area * b.velocity_sup
+                == b.process_flow.properties_in[t].flow_vol_phase["Liq"]
+            )
+
+        @self.Constraint(doc="bed volume based on cylindrical dimensions")
+        def eq_bed_volume(b):
+            return b.bed_volume == b.bed_length * b.bed_area
+
+        @self.Constraint(doc="fluid residence time in the bed")
+        def eq_residence_time(b):
+            return b.residence_time == b.bed_voidage * b.ebct
+
+        @self.Constraint(doc="total mass of gac in the bed")
+        def eq_mass_gac_bed(b):
+            return b.bed_mass_gac == b.particle_dens_bulk * b.bed_volume
+
+        # ---------------------------------------------------------------------
+        # gac CPHSDM intermediate equations
+
         @self.Constraint(
-            doc="Minimum Stanton number to achieve constant pattern solution"
+            doc="minimum Stanton number to achieve constant pattern solution"
         )
-        def eq_min_n_st_cps(b):
+        def eq_min_number_st_cps(b):
             return b.min_N_St == b.a0 * b.N_Bi + b.a1
 
         @self.Constraint(
-            doc="Minimum empty bed contact time to achieve constant pattern solution"
+            doc="minimum empty bed contact time to achieve constant pattern solution"
         )
         def eq_min_ebct_cps(b):
             return b.min_ebct * (1 - b.bed_voidage) * b.kf == b.min_N_St * (
                 b.particle_dia / 2
             )
 
+        @self.Constraint(doc="throughput based on empirical 5-parameter regression")
+        def eq_throughput(b):
+            return b.throughput == b.b0 + b.b1 * (
+                b.conc_ratio_replace**b.b2
+            ) + b.b3 / (1.01 - (b.conc_ratio_replace**b.b4))
+
         @self.Constraint(
-            doc="Minimum fluid residence time in the adsorber bed to achieve a constant pattern solution"
+            doc="minimum fluid residence time in the bed to achieve a constant pattern solution"
         )
-        def eq_min_res_time_cps(b):
-            return b.min_res_time == b.bed_voidage * b.min_ebct
-
-        @self.Constraint(doc="Residence time")
-        def eq_res_time(b):
-            return b.res_time == b.bed_voidage * b.ebct
```

Comment:
Was this where you found your degenerate constraint?

Reply:
There were two issues. One was effectively a circular reference in relating length. volume, cross sectional area, volumetric flowrate, velocity, etc. in a set of equations. The other was an isolated constraint involving the solid mass density of the GAC. Although the DOF would be zero, when providing logical variables to be fixed, the isolated constraint would have no response/ infinite solutions in the optimizer due to 2 unknown, 1 isolated constraint.